weasyl/commishinfo.py

```diff
@@ -20,39 +53,261 @@ def convert_currency(target):
+    except WeasylError:
+        # There was an HTTP error while fetching from the API
+        return None
+
+
+def _charmap_to_currency_code(charmap):
+    """
+    Convert Weasyl's internal single-character representation of currencies
+    to standard ISO4217 codes for use in comparing against exchange rate APIs
+
+    :param charmap: String containing ideally one or zero characters used as currency indicators by Weasyl
+    :return: A 3-letter ISO4217 currency code. Returns "USD" if no match found.
+    """
+    for c in charmap:
+        if c in CURRENCY_CHARMAP:
+            return CURRENCY_CHARMAP.get(c)['code']
+    return "USD"
+
+
+def convert_currency(value, valuecode, targetcode):
+    """
+    Convert between different currencies using the current exchange rate.
+
+    :param value: The amount of currency to be converted
+    :param valuecode: Weasyl-code of the currency $value is in currently
+    :param targetcode: Weasyl-code of the currency $value should be converted to
+    :return: The converted amount
+    """
+    if targetcode == valuecode:
+        return value
+    ratio = currency_ratio(valuecode, targetcode)
+    return value * ratio
+
+
+def currency_ratio(valuecode, targetcode):
+    """
+    Calculate the exchange rate multiplier used to convert from $valuecode to $targetcode
+
+    :param valuecode: Weasyl-code of the currency $value is in currently
+    :param targetcode: Weasyl-code of the currency $value should be converted to
+    :return: The conversion ratio
+    """
+    valuecode = _charmap_to_currency_code(valuecode)
+    targetcode = _charmap_to_currency_code(targetcode)
+    rates = _fetch_rates()
+    if not rates:
+        # in the unlikely event of an error, invalidate our rates
+        # (to try fetching again) and return value unaltered
+        _fetch_rates.invalidate()
+        return 1.0
+    rates = rates["rates"]
+    rates["USD"] = 1.0
+    r1 = rates.get(valuecode)
+    r2 = rates.get(targetcode)
+    if not (r1 and r2):
+        raise WeasylError("Unexpected")
+    return r2 / r1
 
-    content = d.execute("SELECT content FROM commishdesc WHERE userid = %i", [userid], ["element"])
 
+def select_list(userid):
+    query = d.engine.execute("SELECT classid, title, amount_min, amount_max, settings, priceid FROM commishprice"
+                             " WHERE userid = %(userid)s ORDER BY classid, title", userid=userid)
+    classes = d.engine.execute("SELECT classid, title FROM commishclass WHERE userid = %(id)s ORDER BY title", id=userid)
+    content = d.engine.execute("SELECT content FROM commishdesc WHERE userid = %(id)s", id=userid).scalar()
+    preference_tags = d.engine.execute("SELECT DISTINCT tag.title FROM searchtag tag "
+                                       "join artist_preferred_tags pref on pref.tagid = tag.tagid "
+                                       "join login l on l.userid = pref.targetid ", userid=userid).fetchall()
```

Comment:
oh wow I can't believe I didn't notice this in testing. That's definitely an oopsie.